README.md
NETROBOT - Realtime
===================

An experimental fork of [https://github.com/gbinside/netrobots]

Liberamente basato su P-ROBOTS [http://corewar.co.uk/probots/p-robo4.txt]

NOTE: Tests are not yet converted.

Getting Started
===============

Package Requirements
--------------------

On Debian/Ubuntu

  sudo aptitude install python-flask protbuf-compiler libzmq3 libzmq3-dev python-zmq

Starting
--------

On main directory

  python run.py

Open the browser on

  http://localhost:8080/

Launch some demo robot

  cd example/python
  python rabbit.py
  python sniper.py

Robots Coding Instructions
==========================

Programming Language
--------------------

Robots can be written using any programming language, because they communicate with the server using:
* Proto Buffers serialization library
* ZeroMQ message API

API
---

Study:
* "client/netrobots.proto" for a description of the RobotStatus
* "client/connect.py" for the API class
* "example/python" for some example of Robot

WEBSERVER
=========

Il webserver è scritto in Flask e sviluppato in TDD; le post sono codificate come normali POST, quindi ad esempio,
 `speed=100&degree=0` , ma le risposte sono in json.

Per le varie rotte consultare gli esempi o il file `test.py`

BOARD
=====
* L'arena è di 1000x1000 basata in 0,0 in basso a sinistra
* gli angoli si misurano in gradi

              135    90   45
                  \  |  /
                   \ | /
             180 --- x --- 0
                   / | \
                  /  |  \
              225   270   315

* il robot occupa le sue coordinate, con una raggio di 1 (usato per il calcolo delle collisioni)

ROBOT
======

~~I robot sono tutti uguali. Le costanti fisiche sono hardcodate per ora.~~

CANNONATE
=========

I proiettili si intendono in tiro balistico, quindi non vengono considerate le eventuali collisioni inaspettate con robot di passaggio,
 perché non vi sono, inquanto i colpi viaggiano più in alto.
I proiettili sparati fuori dall'arena esplodono fuori, non collidono coi bordi per lo stesso motivo di cui sopra.

NOTE TECNICHE
=============

La versione committata gira con un rapporto temporale di x2

TODO
=====

* ~~I robot hanno tutti le stesse costanti di base, implementare un sistema a punti per personalizzarsi il robot.~~
* Raffinare la logica di sterzo sopra la velocità massima di sterzo (`_max_sterling_speed`), magari con aggiunta di danno autoinflitto
* L'urto con qualcosa infligge 2 punti di danno a prescindere dalla velocità al momento dell'impatto. Farlo dipendere dalla velocità?
* Muri; ora l'arena è vuota, si potrebbere prevedere muri casuali, ma questo imporrebbe modifiche anche allo scanner.


CHANGELOG
2015-03-03

The server now uses for communicating with clients:
* ZeroMQ library, instead of HTTP protocol (clients can be processed in a more fair and controllable way, respect a generic HTTP server)
* ProtoBuf for data conversions, instead of REST (data format is documented, and it is interchangeable between different programming languages, using automatic binding generation)
* REST API is mantained only for the display of the board

client/netrobots.proto contains a description of the accepted format.

client/connect.py is a class for controlling robots. The directory example contains example usages.

The server answer contains a reference to the exact "global-time" of the board simulation ticket.

A Robot reacts to 3 contemporary/optional actions:
* scan
* fire
* drive

The PHP API can send the 3 actions in the same request without needing separate threads.

The server calculate the new board, applying all the changes of robots contemporary. The only constraint it is that the clients send the command within the simulation time-frame.

The server block clients sending too much requests without waiting a response from it. The connect.py API behave always in the correct way.


app/robot/model.py
from math import atan2, degrees, pi, sin, cos, radians
from random import randint
import app

START_COORDS = [(250, 500), (750, 500), (500, 250), (500, 750), (250, 250), (750, 750), (750, 250), (250, 750)]
START_HEADING = [0, 180, 90, 270, 45, 225, 135, 315]


class Robot:
    def __init__(self, board, name, count_of_other=0, configuration=None):

        self.scan_degree = None
        self.scan_resolution = None
        self.scan_distance = None
        self.fired_new_missile = False

        self.last_command_executed_at_global_time = -1

        self._board = board
        self._name = name
        self._max_hit_points = 100
        self._hit_points = self._max_hit_points
        self._winner = False
        self._dead = False
        self._x, self._y = START_COORDS[count_of_other % len(START_COORDS)]
        self._heading = START_HEADING[count_of_other % len(START_HEADING)]
        self._current_speed = 0
        self._required_speed = 0
        self._max_speed = 27  # m/s
        self._acceleration = 9  # m/s^2
        self._decelleration = -5  # m/s^2
        self._max_sterling_speed = 13
        self._max_scan_distance = 700
        self._max_fire_distance = 700
        self._bullet_speed = 500  # m/s
        """
        3%  --    A missile explodes within a 40 meter radius.
        5%  --    A missile explodes within a 20 meter radius.
        10% --    A missile explodes within a 5 meter radius.
        """
        # self._bullet_damage = ((40, 3), (20, 5), (5, 10))
        # changed to progessive damage
        self._bullet_damage = ((40, 3), (20, 2), (5, 5))
        self._reloading = False
        self._reloading_time = 2  # s
        self._reloading_counter = 0.0

        for k, v in configuration.items():
            if v is not None:
                if hasattr(self, '_' + k):
                    setattr(self, '_' + k, v)

        if not app.app.config['TESTING']:
            self._x, self._y = randint(100, 900), randint(100, 900)

        self._board.add_robot(self)

    def calc_value(self):
        return sum([
            self._max_hit_points,
            2 * self._max_speed,
            2 * self._acceleration,
            -2 * self._decelleration,
            2 * self._max_sterling_speed,
            0.01 * self._max_scan_distance,
            0.01 * self._max_fire_distance,
            0.01 * self._bullet_speed,
            0.005 * pi * sum([(x[0] ** 2) * x[1] for x in self._bullet_damage]),
            10 * (3 - self._reloading_time),
        ])

    def get_data(self):
        return dict(
            name=self._name,
            max_hit_points=self._max_hit_points,
            required_speed=self._required_speed,
            max_speed=self._max_speed,
            acceleration=self._acceleration,
            decelleration=self._decelleration,
            max_sterling_speed=self._max_sterling_speed,
            max_scan_distance=self._max_scan_distance,
            max_fire_distance=self._max_fire_distance,
            bullet_speed=self._bullet_speed,
            bullet_damage=self._bullet_damage,
            reloading_time=self._reloading_time,
            reloading_counter=self._reloading_counter,
            fired_new_missile=self.fired_new_missile
        )

    def get_status(self):
        return dict(
            name=self._name,
            hp=self._hit_points,
            heading=self._heading,
            speed=self._current_speed,
            x=self._x,
            y=self._y,
            dead=self._dead,
            winner=self._winner,
            max_speed=self._max_speed,
            reloading=self._reloading,
            is_dead=self.is_dead(),
            scan_degree=self.scan_degree,
            scan_resolution=self.scan_resolution,
            scan_distance=self.scan_distance
        )

    def drive(self, degree, speed):
        if self.is_dead():
            return False
        degree, speed = int(float(degree)) % 360, int(speed)
        if degree != self._heading and self._current_speed > self._max_sterling_speed:
            # overheat
            self.block()
        else:
            self._heading = degree
            self._required_speed = min(speed, self._max_speed)
        return True

    def scan(self, degree, resolution):
        if self.is_dead():
            return False
        degree, resolution = int(degree) % 360, max(1, int(resolution))
        distance = self._board.radar(self, (self._x, self._y), self._max_scan_distance, degree, resolution)

        self.scan_degree = degree
        self.scan_resolution = resolution
        self.scan_distance = distance

        return True

    def no_scan(self):
        self.scan_degree = None
        self.scan_resolution = None
        self.scan_distance = None

        return True

    def cannon(self, degree, distance):
        if self.is_dead():
            self.fired_new_missile = False
            return False
        if self._reloading is False:
            degree, distance = int(float(degree)) % 360, min(int(float(distance)), self._max_fire_distance)
            self._board.spawn_missile((self._x, self._y), degree, distance, self._bullet_speed, self._bullet_damage,
                                      self)
            self._reloading = True
            self._reloading_counter = 0.0

            self.fired_new_missile = True
            return True

        self.fired_new_missile = False
        return False

    def no_cannon(self):
        self.fired_new_missile = False

    def distance(self, xy):
        if isinstance(xy, Robot):
            dist, angle = xy.distance((self._x, self._y))
            angle = (angle + 180) % 360
            return dist, angle
        dx = (xy[0] - self._x)
        dy = (xy[1] - self._y)
        dist = (dx ** 2 + dy ** 2) ** 0.5
        rads = atan2(dy, dx)
        # angle = (360 - degrees(rads)) % 360
        angle = degrees(rads) % 360
        return dist, angle

    def block(self):
        self._current_speed = 0
        self._required_speed = 0

    def tick(self, deltatime):
        # RELOADING
        if self._reloading:
            self._reloading_counter += deltatime
            if self._reloading_counter >= self._reloading_time:
                self._reloading = False
                self._reloading_counter = 0.0

        # SPEED calculation with accelleration/decelleration limit
        if self._current_speed <= self._required_speed:
            delta = accel = min(self._acceleration,
                                float(self._required_speed - self._current_speed) / float(deltatime))
        else:
            delta = decell = max(self._decelleration,
                                 float(self._required_speed - self._current_speed) / float(deltatime))

        # MOVEMENT
        movement = self._current_speed * deltatime + 0.5 * delta * deltatime ** 2
        dx = movement * cos(radians(self._heading))
        dy = movement * sin(radians(self._heading))
        self._x += dx
        self._y += dy

        # UPDATE velocity
        self._current_speed += delta * deltatime

        # COLLISION
        collision = self._board.detect_collision(self, (self._x, self._y), (dx, dy))
        if collision is not None:
            self.block()
            self._x, self._y = collision[:2]
            self.take_damage(collision[2])

    def take_damage(self, hp):
        self._hit_points -= hp
        if self._hit_points <= 0:
            self.block()
            self._hit_points = 0
            self._dead = True

    def get_name(self):
        return self._name

    def get_xy(self):
        return self._x, self._y

    def is_dead(self):
        return self._dead

